docs: add Injectable migration docs (#32581)
angular/angular: aio/content/guide/migration-injectable.md

# Migration to adding missing `@Injectable` decorators

## What does this schematic do?

This schematic adds an `@Injectable()` decorator to a class
if an `@NgModule()` refers to it unless:
  - It already has another decorator such as `@Component()`, `@Directive()` or `@Pipe()`. There is no need to add `@Injectable()` in those cases.
  - The provider definition has `useValue`, `useFactory`, or `useExisting`.

## Why is this migration necessary?

Older versions of Angular, in some cases, allowed injection of
classes without a decorator such as `@Injectable()`. However,
Ivy strictly requires decorators on all injected classes. By strictly

> requiring a decorator, the Ivy compiler can run more quickly

Reviewer: That all sounds like a fancy way of saying "ivy compiles faster" to me :stuck_out_tongue: 

In all serious, it would be fine if you want to link off somewhere to provide an in depth, technical explanation of why it's faster, but at the end of the day "it makes the compiler faster" is as much justification you need in the migration guide

> requiring a decorator, the Ivy compiler can run more quickly

Author: Changed to: 

>Strictly requiring a decorator will enable optimizations in the compiler going forward.

1. Apps that have undecorated base classes that use the following:
    - dependency injection
    - `@Input()` and lifecycle hooks (is this and/or?)
    - that use dependency injection and injection flag, such as `@SkipSelf()`

2. Apps that rely on ancestor class decorators
    - In View Engine, subclasses don't
    have to have an `@Injectable()`
    decorator if they don't have a constructor
    because they implicitly use their superclass’s
    constructor. Though this pattern
    will still work in Angular version 9, it is planned
    for removal in version 10, so you
    should still migrate these apps.

3. A `@Pipe()`/`@Component()`/`@Directive()` that is injected, but not decorated explicitly with `@Injectable()`.

## When should I be adding `@Injectable()` decorators to classes?

<!-- Is this by version 9^^? -->

## Should I update my library?

<!-- what's the time frame here? should I say "continue to work until v10?" -->
Existing published libraries should continue to work until_______.
Beginning with Angular version ____ a missing `@Injectable()`
decorator will throw an error in AOT (in addition to JIT).
Thus, libraries maintainers will need to correctly decorate

> their code as well.

Reviewer 2: ```suggestion
Yes, if your library has any tokens that are meant to be injected, they should be updated with the `@Injectable` annotation.  In a future version of Angular, a missing `@Injectable` decorator will always throw an error.
```

    @NgModule({
      providers: [
        TypeCase,  // TypeCase needs @Injectable()
        {provide: ProvideCase},  // ProvideCase needs @Injectable()
        {provide: ValueCase, useValue: 0}, // No @Injectable() needed because the value will be used
        {provide: FactoryCase, useFactory: ()=> null}, // No @Injectable needed because factory will be used
        {provide: ExistingToken, useExisting: ExistingClass},  // ExistingClass needs @Injectable()
        {provide: ClassToken, useClass: SomeClass},  // SomeClass needs @Injectable()
        PipeCase,   // No @Injectable() needed because it has a @Pipe decorator

      ]
    })

  ```

## Why is this migration necessary?

In our docs, we've always recommended adding `@Injectable()`
decorators to any class that is provided or injected in your application.
However, older versions of Angular did allow injection of a class
without the decorator in certain cases, such as AOT mode.
Ivy strictly requires decorators on all classes that
are provided or injected. Strictly requiring a decorator is
necessary to enable
optimizations in future versions of Angular.

## When should I be adding `@Injectable()` decorators to classes?

> Any class that is provided or injected somewhere must have an `@Injectable()` decorator. The decorator is necessary for the framework to properly create an instance of that class through DI.

Reviewer 2: Oops, figure we should qualify it

```suggestion
Any class that is provided or injected somewhere must have an `@Injectable()` decorator. The decorator is necessary for the framework to properly create an instance of that class through DI.

However, as noted above, classes that already have another class decorator like `@Pipe` do not need both decorators. The existing class decorator will cause the compiler to generate the proper information.
```

# Migration to Add Missing `@Injectable()` Decorators

## What does this schematic do?

This schematic adds an `@Injectable()` decorator to a class
if it's added as a provider unless:

>   - It already has another decorator such as `@Component()`, `@Directive()` or `@Pipe()`. There is no need to add `@Injectable()` in those cases.

Reviewer 2: ```suggestion
  - It already has another decorator such as `@Component()`, `@Directive()` or `@Pipe()`. These decorators already cause the compiler to generate the information we need.
```

  - The provider definition has `useValue`, `useFactory`, or `useExisting`. In

>   these cases, the framework doesn't need the additional creation information

Reviewer 2: Nit: I think we can simplify my earlier suggestion

```suggestion
  these cases, the framework doesn't need the `@Injectable` decorator to create the class because ...
```

  provided by the `@Injectable()` decorator because it can just use the value,
  factory function, or existing instance that was provided.

  ```ts
    // Schematic adds @Injectable()
    class TypeCase{}

    // Schematic adds @Injectable()
    class BaseProviderCase{}

    // Schematic adds @Injectable()
    class ProviderCase extends BaseProviderCase{}

    // Schematic adds @Injectable()
    class ProviderCase2{}

    // Schematic does nothing here.
    class DontNeedCase{}

    // Schematic does nothing here, because class already has decorator
    @Directive()

>     class DirectiveCase{}

Reviewer 2: Let's delete all these definitions and add comments to the providers list below instead

    @NgModule({
      providers: [

>         TypeCase,

Reviewer 2: ```suggestion
        TypeCase,  // TypeCase needs @Injectable
```

>         {provide: ProvideCase},

Reviewer 2: ```suggestion 
        {provide: ProvideCase},    // ProvideCase needs @Injectable
```

>         {provide: DontNeedCase, useValue: 0},

Reviewer 2: Let's disambiguate these `DontNeedCase` tokens, since in real applications, these providers would override each other

```suggestion
        {provide: ValueCase, useValue: 0},   // No @Injectable needed because the value will be used
```

>         {provide: DontNeedCase, useFactory: ()=> null},

Reviewer 2: ```suggestion
        {provide: FactoryCase, useFactory: ()=> null},  // No @Injectable needed because factory will be used
```

>         {provide: DontNeedCase, useExisting: TypeCase},

Reviewer 2: ```suggestion
        {provide: ExistingToken, useExisting: ExistingClass},  // ExistingClass needs @Injectable
```

>         {provide: DontNeedCase, useClass: ProvideCase2},

Reviewer 2: ```suggestion
        {provide: ClassToken, useClass: SomeClass},  // SomeClass needs @Injectable
```

>         DirectiveCase,

Reviewer 2: Let's switch to providing a pipe since it's more realistic (components aren't really provided this way, except maybe in tests).

```suggestion
        PipeCase,   // No @Injectable needed because it has a @Pipe decorator
```

      ]
    })

  ```

## Why is this migration necessary?

> Older versions of Angular allowed injection of

Reviewer 2: I think this is missing a line like:

```suggestion
In our docs, we've always recommended adding `@Injectable` decorators to any class that is provided or injected in your application. However, older versions of Angular did allow injection of a class without the decorator in certain cases, like AOT mode.
```

> classes without an `@Injectable()` decorator in AOT mode. However,

Reviewer 2: ```suggestion
classes without an `@Injectable()` decorator in AOT mode. 
```

Ivy strictly requires decorators on all classes that are provided or

> injected. Strictly requiring a decorator will enable optimizations in the compiler going forward.

Reviewer 2: ```suggestion
injected. Strictly requiring a decorator is necessary to enable optimizations in future versions of Angular.
```

# Migration to Add Missing `@Injectable()` Decorators

## What does this schematic do?

This schematic adds an `@Injectable()` decorator to a class

> if it has been added as a provider anywhere in the application.

Reviewer 2: ```suggestion
if the class has been added as a provider anywhere in the application.
```

> 

Reviewer 2: Sorry, reading through this one more time, I think we might want to move the diff up here, so it's clear that this is the actual schematic change (rather than the comments).

```suggestion
An example diff might look like: 

<Before/After>
```

There are a few cases where the decorator won't be added. For example:

  - It already has another decorator such as `@Component()`, `@Directive()` or `@Pipe()`. These decorators already cause the compiler to generate the necessary information.
  - The provider definition has `useValue`, `useFactory`, or `useExisting`. In
  these cases, the framework doesn't need the `@Injectable()` decorator to create the class because
  because it can just use the value,
  factory function, or existing instance that was provided.

  For example, for the following module definition, the schematic will check
  `TypeCase`, `ProvideCase`, `ExistingClass`, and `SomeClass` to ensure they

>   are marked with the `@Injectable()` decorator.

Reviewer 2: ```suggestion
  are marked with the `@Injectable()` decorator and add one if not.
```

  ```ts
    @NgModule({
      providers: [
        // TypeCase needs @Injectable()
        TypeCase,

>         {provide: ProvideCase},  // ProvideCase needs @Injectable()

Reviewer 2: Oh I meant that we should use that comment style for **each line**, not just the first one. So each provider has a line between it and the comment should be above it (instead of on the right)

```suggestion

        // ProvideCase needs @Injectable()
        {provide: ProvideCase},  
```

        {provide: ValueCase, useValue: 0}, // No @Injectable() needed because the value will be used
        {provide: FactoryCase, useFactory: ()=> null}, // No @Injectable needed because factory will be used
        {provide: ExistingToken, useExisting: ExistingClass},  // ExistingClass needs @Injectable()
        {provide: ClassToken, useClass: SomeClass},  // SomeClass needs @Injectable()
        PipeCase,   // No @Injectable() needed because it has a @Pipe decorator

      ]
    })

  ```

> If the `@Injectable()` decorator was not found, the schematic will add one:

Reviewer 2: We can remove this line and move the diff on top of "There are a few cases"

**Before:**

```ts

> export class TypeCase {}

Reviewer 2: Let's add `...` so it's clear we're yada-yada-ing the contents 

```suggestion
export class TypeCase {...}
```

```

**After:**

```ts
@Injectable()
export class TypeCase {}
```

## Why is this migration necessary?

In our docs, we've always recommended adding `@Injectable()`
decorators to any class that is provided or injected in your application.
However, older versions of Angular did allow injection of a class
without the decorator in certain cases, such as AOT mode.
This means if you accidentally omitted the decorator, your application
may have continued to work despite missing `@Injectable()` decorators in some places.

>  Strictly requiring a decorator is

Reviewer 2: It looks like we're still missing the bottom sentence from this diff:

https://github.com/angular/angular/pull/32581#discussion_r332728065

necessary to enable
optimizations in future versions of Angular.

## When should I be adding `@Injectable()` decorators to classes?

Any class that is provided or injected somewhere must have an `@Injectable()` decorator. The decorator is necessary for the framework to properly create an instance of that class through DI.

However, as noted above, classes that already have another class decorator like `@Pipe` do not need both decorators. The existing class decorator will cause the compiler to generate the proper information.

## Should I update my library?

> Yes, if your library has any tokens that are meant to be injected, they should be updated with the `@Injectable()` annotation.  In a future version of Angular, a missing `@Injectable()` decorator will always throw an error.

Reviewer 2: Whoops, I think this is from my suggestion, but we should be consistent and always say "decorator"

```suggestion
Yes, if your library has any tokens that are meant to be injected, they should be updated with the `@Injectable()` decorator.  In a future version of Angular, a missing `@Injectable()` will always throw an error.
```

  ```ts
    @NgModule({
      providers: [
        // TypeCase needs @Injectable()
        TypeCase,
        // ProvideCase needs @Injectable()
        {provide: ProvideCase},
        // No @Injectable() needed because the value will be used
        {provide: ValueCase, useValue: 0},
        // No @Injectable() needed because factory will be used
        {provide: FactoryCase, useFactory: ()=> null},
        // ExistingClass needs @Injectable()
        {provide: ExistingToken, useExisting: ExistingClass},
        // SomeClass needs @Injectable()
        {provide: ClassToken, useClass: SomeClass},
        // No @Injectable() needed because it has a @Pipe() decorator
        PipeCase,

      ]
    })

  ```

## Why is this migration necessary?

In our docs, we've always recommended adding `@Injectable()`
decorators to any class that is provided or injected in your application.
However, older versions of Angular did allow injection of a class
without the decorator in certain cases, such as AOT mode.
This means if you accidentally omitted the decorator, your application
may have continued to work despite missing `@Injectable()` decorators in some places.

>  Strictly requiring a decorator is

Reviewer 2: Sorry, what I meant in the original comment was that this line should be replaced by the `This is problematic` paragraph.

```suggestion

```
